In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np


# Load cleaned dataset
df = pd.read_csv(
    "../../../data/processed/diabetes_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 229474
Columns : 22


,Diabetes_binary,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,...,AnyHealthcare,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income
0,0.0,1.0,1.0,1.0,40.0,1.0,0.0,0.0,0.0,0.0,...,1.0,0.0,5.0,18.0,15.0,1.0,0.0,9.0,4.0,3.0
1,0.0,0.0,0.0,0.0,25.0,1.0,0.0,0.0,1.0,0.0,...,0.0,1.0,3.0,0.0,0.0,0.0,0.0,7.0,6.0,1.0
2,0.0,1.0,1.0,1.0,28.0,0.0,0.0,0.0,0.0,1.0,...,1.0,1.0,5.0,30.0,30.0,1.0,0.0,9.0,4.0,8.0
3,0.0,1.0,0.0,1.0,27.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,2.0,0.0,0.0,0.0,0.0,11.0,3.0,6.0
4,0.0,1.0,1.0,1.0,24.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,2.0,3.0,0.0,0.0,0.0,11.0,5.0,4.0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================


target_column = "Diabetes_binary"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]



print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")

display(y.head())

FEATURE MATRIX (X)
Rows    : 229474
Columns : 21

TARGET VARIABLE (y)
Rows : 229474


0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: Diabetes_binary, dtype: float64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================


print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True)*100)
    .round(2)
)

TARGET DISTRIBUTION


Diabetes_binary
0.0    194377
1.0     35097
Name: count, dtype: int64


Class Percentage


Diabetes_binary
0.0    84.71
1.0    15.29
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================


print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
HighBP,float64
HighChol,float64
CholCheck,float64
BMI,float64
Smoker,float64
Stroke,float64
HeartDiseaseorAttack,float64
PhysActivity,float64
Fruits,float64
Veggies,float64


In [5]:
# =============================================================================
# STEP 3: TRAIN-TEST SPLIT
# =============================================================================


from sklearn.model_selection import train_test_split


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (183579, 21)
y_train shape : (183579,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (45895, 21)
y_test shape  : (45895,)


In [6]:
# =============================================================================
# CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================


print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True)*100)
    .round(2)
)



print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True)*100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


Diabetes_binary
0.0    84.71
1.0    15.29
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


Diabetes_binary
0.0    84.71
1.0    15.29
Name: proportion, dtype: float64

In [7]:
# =============================================================================
# STEP 4: FEATURE SCALING
# =============================================================================


from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


# Initialize scaler
scaler = StandardScaler()


# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)


# Transform test data using same scaler
X_test_scaled = scaler.transform(X_test)



print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(f"X_train_scaled shape : {X_train_scaled.shape}")
print(f"X_test_scaled shape  : {X_test_scaled.shape}")

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (183579, 21)
X_test_scaled shape  : (45895, 21)


In [8]:
# =============================================================================
# CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns
)



print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(X_train_scaled.head())

SCALED TRAIN DATA PREVIEW


,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,Veggies,...,AnyHealthcare,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income
0,1.096408,-0.889095,0.205176,-0.395435,1.072035,-0.215734,-0.338755,0.601961,-1.256327,0.509271,...,-4.187359,-0.3204,-0.563578,-0.455262,-0.517047,-0.477229,1.13290,-0.350440,0.020709,0.529942
1,1.096408,1.124739,0.205176,0.786628,-0.932805,-0.215734,-0.338755,0.601961,-1.256327,0.509271,...,0.238814,-0.3204,0.375330,-0.455262,-0.517047,-0.477229,-0.88269,0.620283,-0.986375,-1.381726
2,-0.912069,1.124739,0.205176,-0.690951,-0.932805,-0.215734,-0.338755,0.601961,0.795971,0.509271,...,0.238814,-0.3204,-0.563578,-0.455262,-0.517047,-0.477229,-0.88269,0.620283,1.027793,1.007859
3,-0.912069,1.124739,0.205176,-0.986467,-0.932805,-0.215734,-0.338755,0.601961,0.795971,0.509271,...,0.238814,-0.3204,-0.563578,-0.455262,-0.517047,-0.477229,-0.88269,1.591006,0.020709,-0.425892
4,1.096408,-0.889095,0.205176,0.195596,-0.932805,-0.215734,-0.338755,0.601961,0.795971,0.509271,...,0.238814,-0.3204,1.314239,-0.455262,0.698722,2.095428,-0.88269,-0.350440,-0.986375,-0.425892


In [9]:
# =============================================================================
# SCALING VERIFICATION
# =============================================================================


print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.mean().round(2)
)



print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


HighBP                  0.0
HighChol                0.0
CholCheck              -0.0
BMI                    -0.0
Smoker                 -0.0
Stroke                  0.0
HeartDiseaseorAttack    0.0
PhysActivity           -0.0
Fruits                 -0.0
Veggies                -0.0
HvyAlcoholConsump      -0.0
AnyHealthcare           0.0
NoDocbcCost             0.0
GenHlth                 0.0
MentHlth                0.0
PhysHlth                0.0
DiffWalk               -0.0
Sex                    -0.0
Age                     0.0
Education              -0.0
Income                 -0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


HighBP                  1.0
HighChol                1.0
CholCheck               1.0
BMI                     1.0
Smoker                  1.0
Stroke                  1.0
HeartDiseaseorAttack    1.0
PhysActivity            1.0
Fruits                  1.0
Veggies                 1.0
HvyAlcoholConsump       1.0
AnyHealthcare           1.0
NoDocbcCost             1.0
GenHlth                 1.0
MentHlth                1.0
PhysHlth                1.0
DiffWalk                1.0
Sex                     1.0
Age                     1.0
Education               1.0
Income                  1.0
dtype: float64

# BORUTA

## 1. Baseline

In [10]:
!pip install boruta

In [12]:
# =============================================================================
# BORUTA BASELINE FEATURE SELECTION (TOP-K = 5,10,15,20)
# =============================================================================

from boruta import BorutaPy
from sklearn.ensemble import RandomForestClassifier
import pandas as pd
import numpy as np


# =============================================================================
# BORUTA FEATURE SELECTION
# =============================================================================

# Random Forest used internally by Boruta
rf_boruta = RandomForestClassifier(
    n_estimators=200,
    class_weight="balanced",
    random_state=42,
    n_jobs=-1
)


# Initialize Boruta
boruta_selector = BorutaPy(
    estimator=rf_boruta,

    n_estimators="auto",

    verbose=0,

    random_state=42,

    max_iter=100,

    perc=100
)


# -------------------------------------------------------------------------
# Fit Boruta ONLY on training data
# -------------------------------------------------------------------------

boruta_selector.fit(
    X_train_scaled.values,
    y_train.values
)


# =============================================================================
# CREATE BORUTA RANKING
# =============================================================================

boruta_scores = pd.DataFrame({

    "Feature": X_train_scaled.columns,

    "Boruta Rank": boruta_selector.ranking_,

    "Confirmed": boruta_selector.support_,

    "Tentative": boruta_selector.support_weak_

})


# -------------------------------------------------------------------------
# Sort by Boruta rank
# -------------------------------------------------------------------------

boruta_scores = boruta_scores.sort_values(
    by=["Boruta Rank", "Feature"]
).reset_index(drop=True)


print("=" * 70)
print("BORUTA FEATURE RANKING")
print("=" * 70)

print(boruta_scores)


# =============================================================================
# TOP-K FEATURE SELECTION
# =============================================================================

k_values = [5, 10, 15, 20]

boruta_results = {}


for k in k_values:

    # ---------------------------------------------------------------------
    # Select top-k features according to Boruta ranking
    # ---------------------------------------------------------------------

    top_features = (
        boruta_scores
        .head(k)["Feature"]
        .tolist()
    )


    # ---------------------------------------------------------------------
    # Boolean mask for selecting columns
    # ---------------------------------------------------------------------

    mask = X_train_scaled.columns.isin(
        top_features
    )


    # ---------------------------------------------------------------------
    # Select training and testing features
    # ---------------------------------------------------------------------

    X_train_selected = X_train_scaled.loc[
        :, mask
    ]

    X_test_selected = X_test_scaled.loc[
        :, mask
    ]


    # ---------------------------------------------------------------------
    # Store results
    # ---------------------------------------------------------------------

    boruta_results[k] = {

        "features": top_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "scores": boruta_scores.head(k)

    }


    # ---------------------------------------------------------------------
    # Display
    # ---------------------------------------------------------------------

    print("=" * 70)
    print(f"TOP {k} BORUTA FEATURES")
    print("=" * 70)

    print(top_features)

    print("\nBoruta Ranking:")
    print(
        boruta_scores.head(k)
    )


BORUTA FEATURE RANKING
                 Feature  Boruta Rank  Confirmed  Tentative
0                    Age            1       True      False
1                    BMI            1       True      False
2                GenHlth            1       True      False
3                 HighBP            2      False      False
4                 Income            3      False      False
5               PhysHlth            4      False      False
6               HighChol            5      False      False
7              Education            6      False      False
8               MentHlth            7      False      False
9               DiffWalk            8      False      False
10  HeartDiseaseorAttack            9      False      False
11                   Sex           10      False      False
12                Smoker           11      False      False
13                Fruits           12      False      False
14          PhysActivity           13      False      False
15               

In [13]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {
    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,

        eval_metric="logloss",
        scale_pos_weight=(y_train == 0).sum() / (y_train == 1).sum()
    )
}

In [14]:
# =============================================================================
# MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


results = []


for k in k_values:

    Xtr = boruta_results[k]["X_train"]

    Xts = boruta_results[k]["X_test"]

    features = ", ".join(
        boruta_results[k]["features"]
    )


    # ---------------------------------------------------------------------
    # Evaluate every model
    # ---------------------------------------------------------------------

    for model_name, model in models.items():

        # Fit model
        model.fit(
            Xtr,
            y_train
        )


        # Predictions
        y_pred = model.predict(
            Xts
        )


        # Probability predictions
        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        # Store metrics
        results.append({

            "Method": "Boruta",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )

        })


# =============================================================================
# FINAL BORUTA RESULTS TABLE
# =============================================================================

boruta_results_df = pd.DataFrame(
    results
)


display(
    boruta_results_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Boruta,5,Logistic Regression,"Age, BMI, GenHlth, HighBP, Income",0.704347,0.308659,0.752671,0.437787,0.797671
1,Boruta,5,Random Forest,"Age, BMI, GenHlth, HighBP, Income",0.723172,0.300155,0.608349,0.401977,0.736248
2,Boruta,5,XGBoost,"Age, BMI, GenHlth, HighBP, Income",0.695784,0.304544,0.770623,0.436562,0.800745
3,Boruta,10,Logistic Regression,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.709533,0.313077,0.753099,0.442288,0.804871
4,Boruta,10,Random Forest,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.812136,0.340370,0.243482,0.283887,0.740769
5,Boruta,10,XGBoost,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.706765,0.313157,0.768770,0.445031,0.805572
6,Boruta,15,Logistic Regression,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.713150,0.316450,0.754808,0.445941,0.807144
7,Boruta,15,Random Forest,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.837063,0.410317,0.149594,0.219252,0.767797
8,Boruta,15,XGBoost,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.710056,0.315947,0.768913,0.447865,0.808701
9,Boruta,20,Logistic Regression,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.714195,0.318100,0.759652,0.448425,0.810549


In [15]:
boruta_results_df.to_csv(
    "../../../results/diabetes/baseline/boruta_baseline_results.csv",
    index=False
)

## 2. DODA

In [16]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [17]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\Anandha Lakshmi\Downloads\Anaconda\python.exe
DODA: c:\Users\Anandha Lakshmi\Downloads\Anaconda\Lib\site-packages\doda\__init__.py
Selector: c:\Users\Anandha Lakshmi\Downloads\Anaconda\Lib\site-packages\doda\selector.py


In [18]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import HadamardFusion

In [20]:
# =============================================================================
# BORUTA + DODA FEATURE SELECTION
# =============================================================================

import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from boruta import BorutaPy


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "clinical_weights.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = HadamardFusion()


# -------------------------------------------------------------------------
# Store DODA results
# -------------------------------------------------------------------------

doda_results = {}


# =============================================================================
# RUN BORUTA + DODA FOR EACH TOP-K
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"BORUTA + DODA — TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # Random Forest estimator used by Boruta
    # ---------------------------------------------------------------------

    rf_boruta = RandomForestClassifier(
        n_estimators=200,
        class_weight="balanced",
        random_state=42,
        n_jobs=-1
    )


    # ---------------------------------------------------------------------
    # Boruta operator
    # ---------------------------------------------------------------------

    boruta = BorutaPy(
        estimator=rf_boruta,
        n_estimators="auto",
        max_iter=100,
        perc=100,
        random_state=42,
        verbose=0
    )


    # ---------------------------------------------------------------------
    # Wrap Boruta as a DODA operator
    # ---------------------------------------------------------------------

    boruta_operator = SklearnAdapter(
        boruta
    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(
        operators=[boruta_operator],
        provider=provider,
        fusion=fusion,
        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on TRAINING data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(
        X_train_scaled,
        y_train
    )


    # ---------------------------------------------------------------------
    # Transform TEST data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(
        X_test_scaled
    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = selector.get_selected_features()


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store everything needed for later evaluation
    # ---------------------------------------------------------------------

    doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores": selector.raw_math_scores_,

        "math_scores": selector.math_scores_,

        "clinical_weights": selector.clinical_weights_,

        "final_scores": selector.final_scores_
    }


print("\n" + "=" * 70)
print("BORUTA + DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


BORUTA + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x0000014F0B1F5810>
DODA Pipeline Started
Running: BorutaPy
DODA feature order:
['HighBP', 'HighChol', 'CholCheck', 'BMI', 'Smoker', 'Stroke', 'HeartDiseaseorAttack', 'PhysActivity', 'Fruits', 'Veggies']
Number of features: 21
Number of rankings: 21
Operator Scores:
{'BorutaPy_1': {'HighBP': np.float64(0.5), 'HighChol': np.float64(0.2), 'CholCheck': np.float64(0.05555555555555555), 'BMI': np.float64(1.0), 'Smoker': np.float64(0.09090909090909091), 'Stroke': np.float64(0.058823529411764705), 'HeartDiseaseorAttack': np.float64(0.1111111111111111), 'PhysActivity': np.float64(0.07692307692307693), 'Fruits': np.float64(0.08333333333333333), 'Veggies': np.float64(0.07142857142857142), 'HvyAlcoholConsump': np.float64(0.06666666666666667), 'AnyHealthcare': np.float64(0.05263157894736842), 'NoDocbcCost': np.float64(0.0625), 'GenHlth': np.float64(1.0), 'MentHlth': np.f

In [21]:
# CELL 2
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = doda_results[k]["selector"]

# -----------------------------------------------------------------------------
# Raw mathematical scores
# -----------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Normalized mathematical scores
# -----------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Combine both
# -----------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)

math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)

print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
3,BMI,1.000000,1.000000
13,GenHlth,1.000000,1.000000
18,Age,1.000000,1.000000
0,HighBP,0.500000,0.500000
20,Income,0.333333,0.333333
15,PhysHlth,0.250000,0.250000
1,HighChol,0.200000,0.200000
19,Education,0.166667,0.166667
14,MentHlth,0.142857,0.142857
16,DiffWalk,0.125000,0.125000


In [22]:
# CELL 3
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)

clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)

print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
0,HighBP,1.0
3,BMI,1.0
13,GenHlth,1.0
6,HeartDiseaseorAttack,1.0
5,Stroke,1.0
18,Age,0.9
16,DiffWalk,0.9
1,HighChol,0.9
15,PhysHlth,0.8
4,Smoker,0.6


In [23]:
# CELL 4
# =============================================================================
# INSPECT FINAL DODA SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final DODA Score"
    ]
)

final_scores = final_scores.sort_values(
    by="Final DODA Score",
    ascending=False
)

print("=" * 70)
print("FINAL DODA SCORES")
print("=" * 70)

display(final_scores)

FINAL DODA SCORES


,Feature,Final DODA Score
3,BMI,1.000000
13,GenHlth,1.000000
18,Age,0.900000
0,HighBP,0.500000
15,PhysHlth,0.200000
1,HighChol,0.180000
16,DiffWalk,0.112500
6,HeartDiseaseorAttack,0.111111
20,Income,0.100000
14,MentHlth,0.071429


In [24]:
selector.get_selected_features()

['BMI',
 'GenHlth',
 'Age',
 'HighBP',
 'PhysHlth',
 'HighChol',
 'DiffWalk',
 'HeartDiseaseorAttack',
 'Income',
 'MentHlth']

In [25]:
comparison = selector.compare_scores()

comparison


,Feature,Math Rank,Normalized Math Score,Clinical Weight,DODA Rank,Final Score,Rank Change
0,BMI,1,1.0000,1.0,1,1.0000,0
1,GenHlth,2,1.0000,1.0,2,1.0000,0
2,Age,3,1.0000,0.9,3,0.9000,0
3,HighBP,4,0.5000,1.0,4,0.5000,0
4,PhysHlth,6,0.2500,0.8,5,0.2000,1
5,HighChol,7,0.2000,0.9,6,0.1800,1
6,DiffWalk,10,0.1250,0.9,7,0.1125,3
7,HeartDiseaseorAttack,11,0.1111,1.0,8,0.1111,3
8,Income,5,0.3333,0.3,9,0.1000,-4
9,MentHlth,9,0.1429,0.5,10,0.0714,-1


In [26]:
# =============================================================================
# BORUTA + DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


doda_model_results = []


for k in k_values:

    Xtr = doda_results[k]["X_train"]

    Xts = doda_results[k]["X_test"]

    features = ", ".join(
        doda_results[k]["features"]
    )


    # ---------------------------------------------------------------------
    # Evaluate all models
    # ---------------------------------------------------------------------

    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        doda_model_results.append({

            "Method": "Boruta + DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


# =============================================================================
# FINAL RESULTS TABLE
# =============================================================================

boruta_doda_results_df = pd.DataFrame(
    doda_model_results
)


display(
    boruta_doda_results_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Boruta + DODA,5,Logistic Regression,"BMI, GenHlth, Age, HighBP, PhysHlth",0.702887,0.307388,0.752244,0.436436,0.797001
1,Boruta + DODA,5,Random Forest,"BMI, GenHlth, Age, HighBP, PhysHlth",0.728925,0.303608,0.597094,0.402536,0.735492
2,Boruta + DODA,5,XGBoost,"BMI, GenHlth, Age, HighBP, PhysHlth",0.694629,0.303196,0.767773,0.434719,0.800155
3,Boruta + DODA,10,Logistic Regression,"BMI, GenHlth, Age, HighBP, PhysHlth, HighChol,...",0.712583,0.315563,0.752244,0.444613,0.805783
4,Boruta + DODA,10,Random Forest,"BMI, GenHlth, Age, HighBP, PhysHlth, HighChol,...",0.805622,0.338485,0.283944,0.308825,0.731153
5,Boruta + DODA,10,XGBoost,"BMI, GenHlth, Age, HighBP, PhysHlth, HighChol,...",0.708443,0.314562,0.768770,0.446449,0.807820
6,Boruta + DODA,15,Logistic Regression,"BMI, GenHlth, Age, HighBP, PhysHlth, HighChol,...",0.712975,0.316255,0.754523,0.445697,0.807390
7,Boruta + DODA,15,Random Forest,"BMI, GenHlth, Age, HighBP, PhysHlth, HighChol,...",0.829546,0.382111,0.185639,0.249880,0.760447
8,Boruta + DODA,15,XGBoost,"BMI, GenHlth, Age, HighBP, PhysHlth, HighChol,...",0.710753,0.316303,0.767346,0.447956,0.808930
9,Boruta + DODA,20,Logistic Regression,"BMI, GenHlth, Age, HighBP, PhysHlth, HighChol,...",0.714195,0.318100,0.759652,0.448425,0.810549


In [27]:
boruta_doda_results_df.to_csv(
    "../../../results/diabetes/doda/boruta_doda_results.csv",
    index=False
)

In [29]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================

comparison_df = pd.concat(
    [
        boruta_results_df,
        boruta_doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,Boruta,5,Logistic Regression,"Age, BMI, GenHlth, HighBP, Income",0.704347,0.308659,0.752671,0.437787,0.797671
1,Boruta,5,Random Forest,"Age, BMI, GenHlth, HighBP, Income",0.723172,0.300155,0.608349,0.401977,0.736248
2,Boruta,5,XGBoost,"Age, BMI, GenHlth, HighBP, Income",0.695784,0.304544,0.770623,0.436562,0.800745
3,Boruta,10,Logistic Regression,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.709533,0.313077,0.753099,0.442288,0.804871
4,Boruta,10,Random Forest,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.812136,0.340370,0.243482,0.283887,0.740769
5,Boruta,10,XGBoost,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.706765,0.313157,0.768770,0.445031,0.805572
6,Boruta,15,Logistic Regression,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.713150,0.316450,0.754808,0.445941,0.807144
7,Boruta,15,Random Forest,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.837063,0.410317,0.149594,0.219252,0.767797
8,Boruta,15,XGBoost,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.710056,0.315947,0.768913,0.447865,0.808701
9,Boruta,20,Logistic Regression,"Age, BMI, GenHlth, HighBP, Income, PhysHlth, H...",0.714195,0.318100,0.759652,0.448425,0.810549


In [30]:
for k in k_values:

    selector = doda_results[k]["selector"]

    raw_rank = [
        x[0]
        for x in sorted(
            selector.raw_math_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    doda_rank = [
        x[0]
        for x in sorted(
            selector.final_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    print("\n" + "=" * 70)
    print(f"TOP {k}")
    print("=" * 70)

    print("Raw Math Ranking:")
    print(raw_rank)

    print("\nDODA Ranking:")
    print(doda_rank)

    print("\nTop-K Same:",
          raw_rank[:k] == doda_rank[:k])


TOP 5
Raw Math Ranking:
['BMI', 'GenHlth', 'Age', 'HighBP', 'Income', 'PhysHlth', 'HighChol', 'Education', 'MentHlth', 'DiffWalk', 'HeartDiseaseorAttack', 'Sex', 'Smoker', 'Fruits', 'PhysActivity', 'Veggies', 'HvyAlcoholConsump', 'NoDocbcCost', 'Stroke', 'CholCheck', 'AnyHealthcare']

DODA Ranking:
['BMI', 'GenHlth', 'Age', 'HighBP', 'PhysHlth', 'HighChol', 'DiffWalk', 'HeartDiseaseorAttack', 'Income', 'MentHlth', 'Stroke', 'Smoker', 'Education', 'PhysActivity', 'CholCheck', 'Sex', 'Fruits', 'Veggies', 'HvyAlcoholConsump', 'NoDocbcCost', 'AnyHealthcare']

Top-K Same: False

TOP 10
Raw Math Ranking:
['BMI', 'GenHlth', 'Age', 'HighBP', 'Income', 'PhysHlth', 'HighChol', 'Education', 'MentHlth', 'DiffWalk', 'HeartDiseaseorAttack', 'Sex', 'Smoker', 'Fruits', 'PhysActivity', 'Veggies', 'HvyAlcoholConsump', 'NoDocbcCost', 'Stroke', 'CholCheck', 'AnyHealthcare']

DODA Ranking:
['BMI', 'GenHlth', 'Age', 'HighBP', 'PhysHlth', 'HighChol', 'DiffWalk', 'HeartDiseaseorAttack', 'Income', 'MentHlth',

# Interpretation — Boruta + DODA

## 1. Overview

The Boruta baseline and Boruta + DODA experiments were evaluated using four feature-selection sizes:

- Top-5
- Top-10
- Top-15
- Top-20

Each feature subset was evaluated using:

- Logistic Regression
- Random Forest
- XGBoost

The evaluation metrics were:

- Accuracy
- Precision
- Recall
- F1 Score
- ROC-AUC

The purpose of the experiment was to determine whether incorporating clinical knowledge through DODA changes the Boruta-derived feature ranking and whether those changes translate into improved downstream predictive performance.

---

## 2. Boruta Baseline Results

Boruta identified the following features as the highest-ranked features:

### Top-5

1. Age
2. BMI
3. GenHlth
4. HighBP
5. Income

### Top-10

The Top-5 features were followed primarily by:

- PhysHlth
- HighChol
- Education
- MentHlth
- DiffWalk

This indicates that Boruta's data-driven importance ranking placed demographic, metabolic, general-health, and physical-health variables highly.

The strongest Top-5 models were:

| Model | Accuracy | Precision | Recall | F1 | ROC-AUC |
|---|---:|---:|---:|---:|---:|
| Logistic Regression | 0.7043 | 0.3087 | 0.7527 | 0.4378 | 0.7977 |
| Random Forest | 0.7232 | 0.3002 | 0.6083 | 0.4020 | 0.7362 |
| XGBoost | 0.6958 | 0.3045 | 0.7706 | 0.4366 | 0.8007 |

XGBoost achieved the highest recall (0.7706) and ROC-AUC (0.8007) among the Top-5 models.

As the number of features increased, the Random Forest accuracy increased substantially, reaching **0.8424 at Top-20**. However, this was accompanied by a very low recall of **0.1492**, indicating that the high accuracy is largely driven by the majority class.

Therefore, accuracy alone is not an appropriate criterion for selecting the best model in this imbalanced diabetes prediction problem.

---

# 3. Boruta + DODA Results

DODA was subsequently applied to the Boruta ranking using the clinical knowledge weights.

The purpose was not simply to improve prediction, but to modify the data-driven ranking according to clinically informed feature importance.

For Top-5, DODA changed the selected features from:

**Boruta:**

> Age, BMI, GenHlth, HighBP, Income

to:

**Boruta + DODA:**

> BMI, GenHlth, Age, HighBP, PhysHlth

Thus, **PhysHlth replaced Income** in the Top-5 feature set.

This is an important result because it demonstrates that the clinical operator can modify the purely data-driven feature ranking.

---

# 4. Effect of DODA on Feature Ranking

The raw Boruta ranking was:

```text
BMI
GenHlth
Age
HighBP
Income
PhysHlth
HighChol
Education
MentHlth
DiffWalk
HeartDiseaseorAttack
Sex
Smoker
Fruits
PhysActivity
Veggies
HvyAlcoholConsump
NoDocbcCost
Stroke
CholCheck
AnyHealthcare
````

After applying DODA, the ranking became:

```text
BMI
GenHlth
Age
HighBP
PhysHlth
HighChol
DiffWalk
HeartDiseaseorAttack
Income
MentHlth
Stroke
Smoker
Education
PhysActivity
CholCheck
Sex
Fruits
Veggies
HvyAlcoholConsump
NoDocbcCost
AnyHealthcare
```

The major changes occurred in the middle and lower portions of the ranking.

For example:

* `PhysHlth` moved from rank 6 → rank 5
* `HighChol` moved from rank 7 → rank 6
* `DiffWalk` moved from rank 10 → rank 7
* `HeartDiseaseorAttack` moved from rank 11 → rank 8
* `Stroke` moved from rank 19 → rank 11
* `Income` moved from rank 5 → rank 9
* `Education` moved from rank 8 → rank 13

Therefore, DODA is clearly modifying the Boruta ranking rather than simply reproducing it.

---

# 5. Stroke Feature — Important Observation

An important observation concerns the `Stroke` feature.

In the original Boruta ranking:

```text
Stroke → Rank 19
```

The clinical weighting mechanism strongly increased the importance of Stroke. Consequently:

```text
Stroke → Rank 11
```

Therefore, DODA moved Stroke upward by **8 positions**.

However, despite this substantial increase, Stroke **did not enter the Top-10 feature set**.

This is an important finding rather than a failure of the method.

It demonstrates that the clinical weighting operator does not automatically force a clinically important feature into the selected subset. Instead, the final ranking remains dependent on the interaction between:

1. Data-driven Boruta importance
2. Normalized mathematical importance
3. Clinical weights
4. The selected fusion mechanism

Thus, even a strongly weighted clinical feature can remain below the Top-K threshold when its underlying data-driven importance is relatively low.

This provides evidence that DODA acts as a **ranking adjustment mechanism rather than a hard clinical override**.

---

# 6. Top-5 Comparison

The most direct comparison is the Top-5 feature set.

### Boruta

```text
Age
BMI
GenHlth
HighBP
Income
```

### Boruta + DODA

```text
BMI
GenHlth
Age
HighBP
PhysHlth
```

The difference is:

```text
Income   → removed
PhysHlth → introduced
```

The other four features remained common to both approaches.

Therefore, DODA preserved the dominant data-driven features while replacing one feature with a clinically prioritized health-related feature.

---

# 7. Predictive Performance Comparison

## Top-5

| Method        | Model               |     F1 | ROC-AUC |
| ------------- | ------------------- | -----: | ------: |
| Boruta        | Logistic Regression | 0.4378 |  0.7977 |
| Boruta        | Random Forest       | 0.4020 |  0.7362 |
| Boruta        | XGBoost             | 0.4366 |  0.8007 |
| Boruta + DODA | Logistic Regression | 0.4364 |  0.7970 |
| Boruta + DODA | Random Forest       | 0.4025 |  0.7355 |
| Boruta + DODA | XGBoost             | 0.4347 |  0.8002 |

The predictive differences at Top-5 are very small.

This is expected because four of the five selected features remain identical between the two approaches.

Therefore, the main contribution of DODA at this stage is **feature-ranking modification and clinical alignment**, rather than a large immediate improvement in predictive performance.

---

# 8. Top-10 Results

At Top-10, the models continued to show similar behavior.

For Logistic Regression:

* Boruta ROC-AUC = **0.8049**
* Boruta + DODA ROC-AUC = **0.8058**

For XGBoost:

* Boruta ROC-AUC = **0.8056**
* Boruta + DODA ROC-AUC = **0.8078**

This represents a small improvement after incorporating clinical knowledge.

However, the improvement should not be described as a large predictive gain.

Instead, it suggests that DODA can modify the selected feature subset while maintaining approximately comparable predictive performance.

---

# 9. Top-15 Results

At Top-15:

### Logistic Regression

Boruta:

```text
ROC-AUC = 0.8071
F1 = 0.4459
```

Boruta + DODA:

```text
ROC-AUC = 0.8074
F1 = 0.4457
```

### XGBoost

Boruta:

```text
ROC-AUC = 0.8087
F1 = 0.4479
```

Boruta + DODA:

```text
ROC-AUC = 0.8089
F1 = 0.4480
```

Again, the difference is very small.

This indicates that DODA does not substantially degrade predictive performance while changing the ranking toward clinically informed priorities.

---

# 10. Top-20 Results

At Top-20, the predictive performance of Boruta and Boruta + DODA was almost identical.

For XGBoost:

* Boruta ROC-AUC = **0.8116**
* Boruta + DODA ROC-AUC = **0.8116**

For Logistic Regression:

* Boruta ROC-AUC = **0.8105**
* Boruta + DODA ROC-AUC = **0.8105**

Random Forest also showed almost identical performance:

* Boruta ROC-AUC = **0.7767**
* Boruta + DODA ROC-AUC = **0.7782**

Therefore, the clinical ranking adjustment did not produce a meaningful loss of predictive discrimination even when applied to larger feature subsets.

---

# 11. Random Forest Behavior

Random Forest achieved high accuracy at larger Top-K values.

For example:

```text
Boruta Top-20 Random Forest

Accuracy = 0.8424
Precision = 0.4532
Recall = 0.1492
F1 = 0.2245
ROC-AUC = 0.7767
```

However, recall is only approximately **15%**.

This means that although the model correctly classifies many samples overall, it detects relatively few positive diabetes cases.

The same pattern is present after DODA:

```text
Boruta + DODA Top-20 Random Forest

Accuracy = 0.8431
Precision = 0.4598
Recall = 0.1482
F1 = 0.2241
ROC-AUC = 0.7782
```

Therefore, the increase in accuracy should **not** be interpreted as superior clinical performance.

For this imbalanced classification problem, recall, F1-score, ROC-AUC and potentially PR-AUC are more informative than accuracy alone.

---

# 12. Overall Effect of DODA

The results show three main effects.

### 1. DODA changes the feature ranking

The ranking is not identical between Boruta and Boruta + DODA.

For example:

```text
Income:
Rank 5 → Rank 9

PhysHlth:
Rank 6 → Rank 5

DiffWalk:
Rank 10 → Rank 7

HeartDiseaseorAttack:
Rank 11 → Rank 8

Stroke:
Rank 19 → Rank 11
```

This confirms that the clinical knowledge component is actively influencing the ranking.

---

### 2. DODA preserves important data-driven features

The strongest Boruta features remained near the top after DODA:

```text
BMI
GenHlth
Age
HighBP
```

This is desirable because DODA is intended to combine data-driven evidence with clinical knowledge rather than completely override the statistical selector.

---

### 3. Predictive performance remains largely stable

Across Logistic Regression, Random Forest and XGBoost, the differences between Boruta and Boruta + DODA are generally small.

This is an important result.

It suggests that clinical integration can modify feature priorities **without substantially sacrificing predictive discrimination**.

---

# 13. Interpretation of the Stroke Result

The Stroke result is particularly relevant to evaluating the behavior of DODA.

Although Stroke received a high clinical weight, its original Boruta ranking was very low:

```text
Boruta rank = 19
```

After DODA:

```text
DODA rank = 11
```

Therefore:

```text
Rank improvement = 19 - 11 = 8 positions
```

However, it still remained outside Top-10.

This indicates that the DODA framework is **not functioning as a rule-based feature injection mechanism**.

Instead, clinical knowledge modifies the mathematical ranking while allowing the empirical evidence to remain influential.

This is arguably a desirable property for a clinical feature-selection framework because assigning a high clinical weight should not automatically guarantee selection if the feature has weak empirical support in the dataset.

---

# 14. Why Top-K Same = False Is Important

For every tested K:

```text
Top-K Same: False
```

This means that the Top-K feature sets generated by Boruta and DODA are not identical.

Therefore, DODA is producing a meaningful change in the feature-selection output.

However, the difference should be interpreted carefully.

A different feature subset alone does not prove that DODA is superior.

The appropriate interpretation is:

> DODA successfully incorporates clinical knowledge into the data-driven feature ranking, resulting in clinically different feature subsets while maintaining comparable predictive performance.

To establish superiority, additional statistical validation would be required, such as:

* repeated stratified cross-validation
* feature-selection stability analysis
* confidence intervals
* statistical comparison of model performance
* clinical validity assessment

---

# 15. Main Finding

The main finding from the Boruta experiment is **not that DODA dramatically improves accuracy**.

Rather, the experiment demonstrates that:

> **DODA can modify a data-driven Boruta feature ranking using clinically informed weights, promote clinically relevant variables such as Stroke, and produce different Top-K feature subsets while maintaining broadly comparable predictive performance.**

This supports the central motivation of the DODA framework: combining **data-driven feature importance with domain knowledge** rather than relying exclusively on statistical importance.

---

# 16. Research Interpretation

From a research perspective, the Boruta experiment provides evidence for the following hypothesis:

> **A clinically guided feature-selection framework can alter the prioritization of features selected by a purely data-driven method without necessarily compromising downstream predictive performance.**

The diabetes results support this hypothesis.

The strongest evidence is the combination of:

1. **Different feature rankings**
2. **Movement of clinically important features**
3. **Replacement of some data-driven features by clinically prioritized features**
4. **Comparable ROC-AUC and F1 performance**
5. **No substantial deterioration in predictive performance**

However, the current results should be presented as **evidence of feasibility and behavioral effect**, rather than definitive proof that DODA is superior to Boruta.

---

# 17. Important Limitation

The current experiment uses a single train/test split.

Therefore, small differences such as:

```text
ROC-AUC:
0.8056 → 0.8078
```

should not be interpreted as statistically significant improvements.

The next stage should evaluate the methods using repeated or stratified cross-validation and measure:

* Mean ROC-AUC
* Standard deviation
* Mean F1
* Mean recall
* Feature-selection stability
* Jaccard similarity
* Clinical validity

This will determine whether the observed effects of DODA are robust rather than specific to one train/test split.

---

# 18. Final Conclusion

Overall, the Boruta + DODA experiment demonstrates that the DODA framework successfully modifies the feature ranking produced by Boruta.

The most notable example is **Stroke**, which moved from rank **19 to rank 11** after clinical weighting, despite not entering the Top-10 even after an 8-rank improvement.

At the same time, the core high-ranking features such as **BMI, GenHlth, Age and HighBP** remained highly ranked.

Importantly, the resulting feature changes did not cause a substantial deterioration in downstream predictive performance.

Therefore, the experiment supports the interpretation that DODA functions as a **clinical adjustment layer over data-driven feature selection**, rather than as a hard rule-based selector.

The results provide promising evidence that DODA can improve the **clinical alignment and interpretability of feature selection while preserving predictive utility**, but further cross-validation and stability analysis are required before claiming general superiority over Boruta.

```
```
